src/runtime/pkg/katautils/config.go

```diff
@@ -1228,9 +1244,176 @@ func decodeConfig(configPath string) (tomlConfig, string, error) {
+	if err != nil {
+		return tomlConf, resolved, err
+	}
+
 	return tomlConf, resolved, nil
 }
 
+func decodeDropIns(mainConfigPath string, tomlConf *tomlConfig) error {
+	configDir := filepath.Dir(mainConfigPath)
+	dropInDir := filepath.Join(configDir, "config.d")
+
+	files, err := ioutil.ReadDir(dropInDir)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("error reading '%s' directory: %s\n", dropInDir, err)
+		} else {
+			return nil
+		}
+	}
+
+	for _, file := range files {
+		dropInFpath := filepath.Join(dropInDir, file.Name())
+		updateFromDropIn(dropInFpath, tomlConf)
+	}
+
+	return nil
+}
+
+func updateFromDropIn(dropInFpath string, tomlConf *tomlConfig) error {
+	configData, err := os.ReadFile(dropInFpath)
+	if err != nil {
+		return fmt.Errorf("error reading file %s: %s\n", dropInFpath, err)
+	}
+
+	// Ordinarily, BurntSushi only updates fields of tomlConfig that are
+	// changed by the file and leaves the rest alone.  This doesn't apply
+	// though to tomlConfig substructures that are stored in maps.  Their
+	// previous contents are erased by toml.Decode() and only fields changed by
+	// the file are set.  To work around this, a bit of juggling is needed to
+	// preserve the previous contents and merge them manually with the incoming
+	// changes afterwards, using reflection.
+	tomlConfOrig := tomlConf.Clone()
+
+	var md toml.MetaData
+	md, err = toml.Decode(string(configData), &tomlConf)
+
+	if err != nil {
+		return fmt.Errorf("error decoding file: %s\n", err)
+	}
+
+	if len(md.Undecoded()) > 0 {
+		msg := fmt.Sprintf("warning: undecoded keys in %s: %+v\n", dropInFpath, md.Undecoded())
+		kataUtilsLogger.Warn(msg)
+	}
+
+	for _, key := range md.Keys() {
+		err = applyKey(*tomlConf, key, &tomlConfOrig)
+		if err != nil {
+			msg := fmt.Sprintf("error applying key '%+v' from drop-in file '%s'\n", key, dropInFpath)
+			kataUtilsLogger.Warn(msg)
+		}
+	}
+
+	tomlConf.Hypervisor = tomlConfOrig.Hypervisor
+	tomlConf.Agent = tomlConfOrig.Agent
+
+	return nil
+}
+
+func applyKey(sourceConf tomlConfig, key []string, targetConf *tomlConfig) error {
+	// Any key that might need treatment provided by this function has to have
+	// (at least) three components: [ map_name map_key_name field_toml_tag ],
+	// e.g. [agent kata enable_tracing] or [hypervisor qemu confidential_guest].
+	if len(key) < 3 {
+		return nil
+	}
+	switch key[0] {
+	case "agent":
+		return applyAgentKey(sourceConf, key[1:], targetConf)
+	case "hypervisor":
+		return applyHypervisorKey(sourceConf, key[1:], targetConf)
+		// The table the 'key' is in is not stored in a map so no special handling
+		// is needed.
+	}
+	return nil
+}
+
+// Both of the following functions copy the value of a 'sourceConf' field
+// identified by the TOML tag in 'key' into the corresponding field in
+// 'targetConf'.
+func applyAgentKey(sourceConf tomlConfig, key []string, targetConf *tomlConfig) error {
+	agentName := key[0]
+	tomlKeyName := key[1]
```

Comment:
The `len` check is performed by the caller (see line 1321 in `applyKey()`).  I moved the check there as it results in more compact code, ensures that whatever is called by `applyKey()` (now or in the future) is guaranteed to have the length already checked, and it's also a tiny optimisation as well (we exclude keys that don't require additional handling as early as possible).

Reply:
I see that, but we try to ensure that all functions are coded defensively. Yes, the code works now, but the caller might be reworked in the future and silently break. Speaking of which, please add unit tests for all the new functions you've added on this PR.

Reply:
James, thanks for your comments!

As for testing: the only function that this PR adds is really just `decodeDropIns()`, the rest are just details of its implementation. I know this is not very obvious but I couldn't find a way within the context I was working in to make it more explicit. Ideally, this PR would be a package with `decodeDropIns()` becoming `DecodeDropIns()` and the rest  private but seeing as the whole configuration handling is in `katautils` (not even `config` or similar), I thought adding a package for my changes wouldn't be very fitting.

In my experience, testing private implementation details is generally not recommended as it makes for brittle tests that don't add much if anything at all as they wouldn't in this case - I believe that testing `decodeDropIns()` fully exercises the private implementation functions as well.  Of course I'm happy to add more tests of `decodeDropIns()` though if deemed necessary!

As for moving the `len` checks to the lower-level functions, I really don't think this would add to defensive coding.  `applyAgentKey()` and `applyHypervisorKey()` are just helpers of `applyKey()`, not API functions that could be called from anywhere. They would actually be quite hard to call correctly from anywhere but `applyKey()`, and it's not clear why anyone would even want to do that as their purpose is very narrow and tailored to what a single specific caller needs.

Apart from that, I believe the most likely future refactoring would be if another member of `tomlConfig` needs the special treatment `applyKey()` provides. In other words, if there's another member of `tomlConfig` which is a `map` - `runtime` comes to mind (?).  In that case, `applyXyzKey()` would need to be added and its implementation would benefit from being safe _by default_ from out-of-bounds key array accesses instead of having to keep in mind that it has to check bounds.

I'll make the requested changes if you insist, this is just to let you know why I wouldn't necessarily consider them beneficial.